### PR TITLE
Add Invoke Statements to LiveRangeAnalysis

### PR DESCRIPTION
**Pull request by calebmkim**
Adds Invoke statements as possible nodes in the Live Range Analysis, not just groups. The reason why we need to do this is the following: say we have the following program, and assume the component `dbg` has been marked as state shareable. 
```
component main() -> () {
  cells {
    d1 = dbg();
    d2 = dbg();
  }
  wires {
  }
  control {
    par{
       invoke d1()();
       invoke d2()();
    }
  }
}
```
As currently constructed, since `dbg` is marked as `state_shareable`, the Live Range Analysis will share d1 and d2 So we need to add Invoke statements as nodes of the Live Range Analysis too. 

This did not seem to be too much more work, although I am not as familiar w/ the codebase for a lot of the analysis stuff, so I may have not been understanding this correctly. 

- I added share set (not state_shareable) as a field of LiveRangeAnalysis since it would be easier to add share cells as active in the invoke stmts that use them in the `build_live_ranges` method. 

Once we get this working correctly, then we will be able to do the infer-sharing stuff.

### CODE REVIEW

**Comment by rachitnigam** [comment]
This looks cool! I think we should do two other things in this PR:
1. In `default_passes.rs`, move the cell-sharing pass to run before `compile-invoke` if it doesn't already.
2. Remove the variable detection stuff from the sharing pass because that is only needed to treat a group as an invoke
3. Run the `group2invoke` pass before the cell sharing pass so that the places where variable detection did its magic, now this invoke handling logic will just work

**Comment by rachitnigam** [comment]
Ah, I see. Group2invoke is more conservative than we’d like in this case. It is possible to turn the group above into an invoke but you’d have to use the invoke-with syntax to run the assignments to the gt while the invoke is active. I guess we can keep variable detection for now and open an issue saying that this case needs to be handled by group2invoke before we can remove variable detection.

an example of invoke-with syntax: https://github.com/cucapra/calyx/blob/118842e67913d6061cc52447f928d5c9abb6aacb/tests/correctness/invoke-with.futil#L26

File: tests/passes/cell-share/par-invoke-adders.futil [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  control {
+    seq{
+      par{
+       invoke d1(in= 32'd2)(out1 = add1.left, out2= add1.right);
+       invoke d2(in= add1.out)(out1 = add2.left, out2= add2.right);
+      }
+      par{
+       invoke d2(in= add1.out)(out1 = add2.left, out2= add2.right);
+      }
```
Comment (rachitnigam) [inline tests/passes/cell-share/par-invoke-adders.futil:40]: Remove the `par` and the newlines

**Review by rachitnigam — APPROVED** [review]
LGTM!

**Comment by sampsyo** [comment]
Indeed, pretty rad!

**Pull request merged**